lib/signet/sleuth.ex

```diff
@@ -50,77 +57,104 @@ defmodule Signet.Sleuth do
     end
   end
 
-  defp try_decode(query_res, selector) do
+  defp try_decode(query_res, selector, decode_structs) do
     try do
-      {:ok, ABI.decode(%ABI.FunctionSelector{types: selector.returns}, query_res)}
+      {:ok,
+       ABI.decode(
+         %ABI.FunctionSelector{types: selector.returns},
+         query_res,
+         decode_structs: decode_structs
+       )}
     rescue
       e ->
         {:error, "error decoding: #{inspect(e)}"}
     end
   end
 
-  defp encode_map(res, types, annotated, decode_binaries) do
-    Enum.map(Enum.zip(res, Enum.with_index(types)), fn {res, {type, i}} ->
-      encode_item(res, type, i, annotated, decode_binaries)
+  defp postprocess(results, named_types, opts) when is_list(results) and is_list(named_types) do
+    be_obvious = Keyword.get(opts, :be_obvious, false)
+
+    results
+    |> Enum.zip(named_types)
+    |> Enum.map(fn {it, t} -> {t.name, postprocess(it, t.type, opts)} end)
```

Comment:
this function would actually be just these three lines if it wasn't for `when not be_obvious`, fwiw.